cli/tools/clean.rs

```diff
@@ -69,6 +92,502 @@ pub fn clean(flags: Arc<Flags>) -> Result<(), AnyError> {
+  let deno_dir_root_canonical =
+    try_get_canonicalized_root_dir(&sys, &deno_dir.root)
+      .unwrap_or(deno_dir.root.clone());
+
+  let mut permit = main_graph_container.acquire_update_permit().await;
+  let graph = permit.graph_mut();
+  graph.packages = PackageSpecifiers::default();
+  let graph_builder = factory.module_graph_builder().await?;
+  graph_builder
+    .build_graph_with_npm_resolution(
+      graph,
+      CreateGraphOptions {
+        // loader: Some(&mut NoLoader),
+        loader: None,
+        graph_kind: graph.graph_kind(),
+        is_dynamic: false,
+        roots: roots.clone(),
+        npm_caching: crate::graph_util::NpmCachingStrategy::Manual,
+      },
+    )
+    .await?;
+
+  let npm_resolver = factory.npm_resolver().await?;
+
+  let mut keep = HashSet::new();
+  let mut npm_reqs = Vec::new();
+
+  let mut keep_paths_trie = PathTrie::new();
+  if deno_dir_root_canonical != deno_dir.root {
+    keep_paths_trie
+      .add_rewrite(deno_dir.root.clone(), deno_dir_root_canonical.clone());
+  }
+  for (_, entry) in graph.walk(
+    roots.iter(),
+    deno_graph::WalkOptions {
+      check_js: deno_graph::CheckJsOption::False,
+      follow_dynamic: true,
+      kind: graph.graph_kind(),
+      prefer_fast_check_graph: false,
+    },
+  ) {
+    match entry {
+      deno_graph::ModuleEntryRef::Module(module) => match module {
+        deno_graph::Module::Js(js_module) => {
+          keep.insert(&js_module.specifier);
+        }
+        deno_graph::Module::Json(json_module) => {
+          keep.insert(&json_module.specifier);
+        }
+        deno_graph::Module::Wasm(wasm_module) => {
+          keep.insert(&wasm_module.specifier);
+        }
+        deno_graph::Module::Npm(npm_module) => {
+          if let Some(managed) = npm_resolver.as_managed() {
+            let id = managed
+              .resolution()
+              .resolve_pkg_id_from_deno_module(npm_module.nv_reference.nv())
+              .unwrap();
+            npm_reqs
+              .extend(managed.resolution().resolve_pkg_reqs_from_pkg_id(&id));
+          }
+        }
+        deno_graph::Module::Node(_) => {}
+        deno_graph::Module::External(_) => {}
+      },
+      deno_graph::ModuleEntryRef::Err(_) => {}
+      deno_graph::ModuleEntryRef::Redirect(_) => {}
+    }
+  }
+
+  for url in &keep {
+    if url.scheme() == "http" || url.scheme() == "https" {
+      if let Ok(path) = http_cache.local_path_for_url(url) {
+        keep_paths_trie.insert(&path);
+      }
+    }
+    if let Some(path) = deno_dir
+      .gen_cache
+      .get_cache_filename_with_extension(url, "js")
+    {
+      let path = deno_dir.gen_cache.location.join(path);
+      keep_paths_trie.insert(&path);
+    }
+  }
+
+  let npm_cache = factory.npm_cache()?;
+  let snap = npm_resolver.as_managed().unwrap().resolution().snapshot();
+  // TODO(nathanwhit): remove once we don't need packuments for creating the snapshot from lockfile
```

Comment:
Isn't this the case with lockfile v5?

Reply:
yes, but we don't know if it's actually lockfile v5. they could still be using v4

Reply:
Ah, good point 👍 let's keep it for now